sdk/dotnet/Pulumi.Automation/Events/Events.cs

```diff
@@ -0,0 +1,495 @@
+            IDictionary<OperationType, int> resourceChanges,
+            IDictionary<string, string> policyPacks)
+        {
+            MaybeCorrupt = maybeCorrupt;
+            DurationSeconds = durationSeconds;
+            ResourceChanges = resourceChanges.ToImmutableDictionary();
+            PolicyPacks = policyPacks.ToImmutableDictionary();
+        }
+    }
+
+    /// <summary>
+    /// DiffKind describes the kind of a particular property diff.
+    /// </summary>
+    public enum DiffKind
+    {
+        /// <summary>
+        /// Add indicates that the property was added.
+        /// </summary>
+        Add,
+        /// <summary>
+        /// AddReplace indicates that the property was added and requires that the resource be replaced.
+        /// </summary>
+        AddReplace,
+        /// <summary>
+        /// Delete indicates that the property was deleted.
+        /// </summary>
+        Delete,
+        /// <summary>
+        /// DeleteReplace indicates that the property was deleted and requires that the resource be replaced.
+        /// </summary>
+        DeleteReplace,
+        /// <summary>
+        /// Update indicates that the property was updated.
+        /// </summary>
+        Update,
+        /// <summary>
+        /// UpdateReplace indicates that the property was updated and requires that the resource be replaced.
+        /// </summary>
+        UpdateReplace,
+    }
+
+    /// <summary>
+    /// PropertyDiff describes the difference between a single property's old and new values.
+    /// </summary>
+    public class PropertyDiff
+    {
+        /// <summary>
+        /// Kind is the kind of difference.
+        /// </summary>
+        public DiffKind Kind { get; }
+
+        /// <summary>
+        /// InputDiff is true if this is a difference between old and new inputs rather than old state and new inputs.
+        /// </summary>
+        public bool InputDiff { get; }
+
+        internal PropertyDiff(DiffKind kind, bool inputDiff)
+        {
+            Kind = kind;
+            InputDiff = inputDiff;
+        }
+    }
+
+    /// <summary>
+    /// StepEventMetadata describes a "step" within the Pulumi engine, which is any concrete action
+    /// to migrate a set of cloud resources from one state to another.
+    /// </summary>
+    public class StepEventMetadata
+    {
+        /// <summary>
+        /// Op is the operation being performed.
+        /// </summary>
+        public OperationType Op { get; }
+
+        public string Urn { get; }
+
+        public string Type { get; }
+
+        /// <summary>
+        /// Old is the state of the resource before performing the step.
+        /// </summary>
+        public StepEventStateMetadata? Old { get; }
+
+        /// <summary>
+        /// New is the state of the resource after performing the step.
+        /// </summary>
+        // TODO: can this actually ever be null?
+        public StepEventStateMetadata? New { get; }
```

Comment:
Can `StepEventMetadata.New` ever be null?

Reply:
Yes, on `Destroy` operations. Just like `Old` is `null` on `Create` operations.